test/extended/apiserver/apply.go

```diff
@@ -0,0 +1,168 @@
+package apiserver
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	g "github.com/onsi/ginkgo"
+	o "github.com/onsi/gomega"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	discocache "k8s.io/client-go/discovery/cached"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/restmapper"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	exetcd "github.com/openshift/origin/test/extended/etcd"
+	exutil "github.com/openshift/origin/test/extended/util"
+)
+
+// TODO(marun) Evaluate the CRDs that ocp with (~70 at time of
+// writing) which will involve creating the associated stubs.
+var _ = g.Describe("[sig-api-machinery][Feature:ServerSideApply] Server-Side Apply", func() {
+	var (
+		// mapper is used to translate the gvr provided by etcd
+		// storage data to the gvk required to create correct resource
+		// yaml.
+		mapper *restmapper.DeferredDiscoveryRESTMapper
+	)
+
+	defer g.GinkgoRecover()
+
+	// Defer project creation to tests that require it by calling
+	// NewCLIWithoutNamespace instead of NewCLI.
+	oc := exutil.NewCLIWithoutNamespace("server-side-apply")
+
+	g.BeforeEach(func() {
+		// Only init once per worker
+		if mapper != nil {
+			return
+		}
+		kubeClient, err := kubernetes.NewForConfig(oc.AdminConfig())
+		o.Expect(err).NotTo(o.HaveOccurred())
+		mapper = restmapper.NewDeferredDiscoveryRESTMapper(discocache.NewMemCacheClient(kubeClient.Discovery()))
+	})
+	storageData := exetcd.OpenshiftEtcdStorageData
+	for key := range storageData {
+		gvr := key
+		data := storageData[gvr]
+
+		// Apply for core types is already well-tested, so skip
+		// openshift types that are just aliases.
+		aliasToCoreType := data.ExpectedGVK != nil
+		if aliasToCoreType {
+			continue
+		}
+
+		g.It(fmt.Sprintf("should work for %s", gvr), func() {
+			// Create at most one namespace only if needed.
+			var testNamespace string
+
+			for _, prerequisite := range data.Prerequisites {
+				resourceClient, unstructuredObj, namespace := createResource(oc, mapper, prerequisite.GvrData, prerequisite.Stub, testNamespace)
+				testNamespace = namespace
+				defer deleteResource(resourceClient, unstructuredObj.GetName())
+			}
+
+			resourceClient, unstructuredObj, namespace := createResource(oc, mapper, gvr, data.Stub, testNamespace)
+			testNamespace = namespace
+			defer deleteResource(resourceClient, unstructuredObj.GetName())
+
+			serializedObj, err := json.Marshal(unstructuredObj.Object)
+			o.Expect(err).NotTo(o.HaveOccurred())
+
+			g.By(fmt.Sprintf("updating the %s via apply", unstructuredObj.GetKind()))
+			obj, err := resourceClient.Patch(
+				context.Background(),
+				unstructuredObj.GetName(),
+				types.ApplyPatchType,
+				serializedObj,
+				metav1.PatchOptions{
+					FieldManager: "apply_test",
```

Comment:
is this what it takes to do server-side apply?

Reply:
the patch type especially.

Reply:
oh, okay, it points to the client that made the update.

Reply:
This just points to who made an update, and it can be supplied with CREATE, UPDATE or PATCH. The key to performing server-side apply is `types.ApplyPatchType`.